### PR TITLE
CSHARP-4335: Do not connect to mongocryptd if shared library is loaded.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 1]
```diff
@@ -250,6 +249,75 @@ public ClientEncryptionProseTests(ITestOutputHelper testOutputHelper)
...
+                    var coll = GetCollection(clientEncrypted, __collCollectionNamespace);
+
+                    listenerThread.Start(tcpListener);
+
+                    _ = Record.Exception(() => Insert(coll, async, new BsonDocument("unencrypted", "test")));
+
+                    if (listenerThread.Join(timeout))
+                    {
+                        // This exception is never thrown when mognocryptd mongoClient is not spawned which is expected behavior.
+                        // However, if we intentionally break that logic to spawn mongocryptd mongoClient regardless shared library,
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:290]:
```suggestion
                        // However, if we intentionally break that logic to spawn mongocryptd mongoClient regardless of shared library,
```

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 2]
```diff
@@ -250,6 +249,75 @@ public ClientEncryptionProseTests(ITestOutputHelper testOutputHelper)
...
+                        // this exception sometimes won't be thrown. In all such cases the spent time in listenerThread.Join is higher
+                        // or really close to timeout. So it's unclear why Join doesn't throw in that cases, but that logic is unrelated
+                        // to the driver and csfle in particular. We rely on the fact that even if we break this logic,
+                        // we run this test more than 1 times
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:294]:
```suggestion
                        // we run this test more than once.
```

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 3]
```diff
@@ -250,6 +249,75 @@ public ClientEncryptionProseTests(ITestOutputHelper testOutputHelper)
...
+
+            void ThreadStart(object param)
+            {
+                try
+                {
+                    var tcpListener = (TcpListener)param;
+                    tcpListener.Start();
+                    using var client = tcpListener.AcceptTcpClient();
+                    // Perform a blocking call to accept requests.
+                    // if we're here, than something queries port 27030
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:312]:
```suggestion
                    // if we're here, then something queries port 27030.
```

**Review by BorisDog — CHANGES_REQUESTED** [review]

**Comment by DmitryLukyanov** [comment]
ready for review

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request merged**